ARCs/arc-0014.md

```diff
@@ -0,0 +1,246 @@
+- The secret keys are stored into a wallet;
+- Users won’t use different *PKa* for authentication;
+- Users won’t change their public address for authentication;
+- Users **MUST** use Algorand compliant keys to sign the messages;
+- Multisig and LogicSig are not supported;
+- Users **MUST NOT** use rekeyed accounts.
+
+> Rekey breaks ARC-14 security. Applications with high security requirements **MUST** check that Users' accounts have not been rekeyed.
+
+### Authentication Mechanism
+
+The authentication mechanism defined in this ARC works as follows: a User communicates to the Verifier her intention to authenticate with her Algorand account <*PKa, SKa*>.
+
+> Algorand transforms traditional 32-bytes cryptographic keys into more readable and user-friendly objects. A detailed description of such a transformation can be found in the <a href="https://developer.algorand.org/docs/get-details/accounts/#keys-and-addresses">developer portal</a>.
+
+The Verifier requests the signing of a random message is required to be signed with the secret key *SKa* associated with the User's Algorand account. The User generates the digital signature using a Wallet, and sends it back to the Verifier. At that stage, the Wallet **MUST** check that the User is signing the message for the correct Verifier (to protect Users against man-in-the-middle attack). Finally, the Verifier verifies the signature and, if everything is good, authenticates the User.
+
+```mermaid
+sequenceDiagram
+    actor A as Alice
+    participant W as Wallet
+    actor B as Bob
+    A-->>W: Connect to Bob with PKa
+    activate W
+    W->>B: GET message for PKa
+    B->>B: Create a new msg for PKa
+    B->>W: msg
+    Note right of A: Confirm signature
+    A-->>W: 
+    W->>W: Check <msg, Bob>
+    W->>B: <PKa, Sig(msg)>
+    deactivate W
+    B->>B: Verify Sig(msg) with PKa and msg
+    Note right of B: if Sig(msg) is valid<br/>then authenticate user
+    
+```
+
+The diagram above summarizes the mechanism. We consider a User, **Alice**, owner of an Algorand account <*PKa, SKa*> of which the secret key *SKa* is stored into a **wallet**.
+
+> A wallet is any type of Algorand wallet, such as hot wallets like <a href="https://www.purestake.com/technology/algosigner/">AlgoSigner</a>, <a href="https://wallet.myalgo.com/">MyAlgo Wallet</a> for browser and mobile wallets used through <a href="https://developer.algorand.org/docs/get-details/walletconnect/">WalletConnect</a>, and cold wallets like the <a href="https://www.ledger.com">Ledger Nano</a>.
+
+Alice authenticates herself to a Validator, **Bob**, providing the digital signature `Sig(msg)` of a random message `msg` provided by Bob. The mechanism proceeds as follows:
+
+1. Alice sends an authentication request to Bob via her Algorand wallet using *PKa*;
+2. Bob generates a message `msg` for *PKa* and request Alice to sign it;
+3. Alice signs `msg` using her wallet which must check that `msg` is for Bob;
```

Comment:
What's the mechanism for the wallet to verify that `msg` comes from Bob, i.e. the `Validator`?

Reply:
Wallets should check the `service` param of the `Authentication Message` to be signed.

Reply:
Solving in favour of similar thread below.